File: 2022/day12/solve1-v2.exs
defmodule Main do
  def go() do
    # "input-small.txt"
    graph =
      "input-large.txt"
      |> File.read!()
      |> String.split("\n", trim: true)
      |> Enum.map(&String.codepoints(&1))
      |> Enum.with_index()
      |> Enum.map(fn {row, i} ->
        row
        |> Enum.with_index()
        |> Enum.map(fn {item, j} -> to_node(item, i, j) end)
      end)

    graph
    |> List.flatten()
    |> Enum.filter(&Map.get(&1, :is_start))
    |> bfs(graph)
  end

  def key(node), do: "#{node.i},#{node.j}"

  def bfs(nodes, graph) do
    distances = %{key(hd(nodes)) => 0}
    bfs(nodes, graph, %{}, distances)
  end

  def bfs(nodes, graph, visited, distances) do
    [node | tail] = nodes

    cond do
      node[:is_end] ->
        distances[key(node)]

      !Map.get(visited, key(node)) ->
        visited = Map.put(visited, key(node), true)
        neighbors = get_neighbors(node, graph)

        distances =
          neighbors
          |> Enum.into(distances, &{key(&1), distances[key(node)] + 1})

        (tail ++ neighbors) |> bfs(graph, visited, distances)

      true ->
        bfs(tail, graph, visited, distances)
    end
  end

  def p(o) do
    IO.inspect(o)
  end

  def get_neighbors(node, graph) do
    [{node.i, node.j + 1}, {node.i, node.j - 1}, {node.i - 1, node.j}, {node.i + 1, node.j}]
    |> Enum.map(&get_at(graph, &1))
    |> Enum.filter(&neighbor(node, &1))
  end

  def neighbor(_node1, nil), do: nil

  def neighbor(node1, node2) do
    if node1.h >= node2.h - 1, do: node2, else: nil
  end

  def get_at(_graph, {_, -1}), do: nil
  def get_at(_graph, {-1, _}), do: nil
  def get_at(graph, {i, j}), do: get_at(Enum.at(graph, i), j)
  def get_at(nil, _j), do: nil
  def get_at(row, j), do: Enum.at(row, j)

  def to_node("S", i, j) do
    to_node("a", i, j)
    |> Map.merge(%{is_start: true, distance: 0, mark: "S"})
  end

  def to_node("E", i, j) do
    to_node("z", i, j)
    |> Map.merge(%{is_end: true, mark: "E"})
  end

  def to_node(c, i, j) do
    %{mark: c, h: c |> to_charlist() |> hd(), i: i, j: j}
  end
end

IO.inspect(Main.go())

# 31 - input-small.txt answer
# 383 - input-large.txt answer

File: 2022/day12/solve1-v2.pl
#!/usr/bin/env perl
# Recursive BFS algorithm.
use v5.20;
use warnings;
use subs qw(p);

use List::Util qw(sum0);
use Data::Dumper qw(Dumper);
$Data::Dumper::Sortkeys = 1;
$Data::Dumper::Terse = 1;
$Data::Dumper::Indent = 1;

#open my $file, '<', 'input-small.txt';
open my $file, '<', 'input-large.txt';
my @lines = <$file>;
chomp @lines;
my $marks = [ map [ split // ], @lines ];
my $max_x = @{$marks->[0]} - 1;
my $max_y = @$marks - 1;
my $graph = {};

for (my $x = 0; $x <= $max_x; $x++) {
    for (my $y = 0; $y <= $max_y; $y++) {
        my $key = key($x, $y);
        my $node = $graph->{$key} = { x => $x, y => $y, key => $key };
        my $mark = $marks->[$y][$x];
        $node->{mark} = $mark;
        $node->{h} = ord $mark;
        if ($mark eq 'S') {
            $node->{h} = ord 'a';
            $node->{is_start} = 1;
            $node->{distance} = 0;
        } elsif ($mark eq 'E') {
            $node->{h} = ord 'z';
            $node->{is_end} = 1;
        }
    }
}

my ($start) = grep { $_->{is_start} } values %$graph;
bfs([$start]);

sub bfs {
    my ($nodes) = @_;
    my $node = shift @$nodes;
    return unless $node;
    #say "$node->{key} d:$node->{distance} v:" . ($node->{v} ? 1 : 0);
    if (not $node->{v}) {
        if ($node->{is_end}) { say "best distance: $node->{distance}" and exit }
        my @children = kids($graph, $node);
        for my $child (@children) { $child->{distance} //= $node->{distance} + 1 }
        push @$nodes, @children;
        #p [ map $_->{key}, @$nodes ];
    }
    $node->{v} = 1;
    @_ = ($nodes);
    goto &bfs;
    #bfs($nodes);
}

sub kids {
    my ($graph, $node) = @_;
    my ($x, $y) = @{$node}{qw(x y)};
    return (
        grep { not $_->{v} }
        grep { $_ and $node->{h} >= ($_->{h} - 1) }
        map { $graph->{key(@$_)} }
        [$x, $y+1], [$x, $y-1], [$x+1, $y], [$x-1, $y]
    );
}

sub key { "$_[0],$_[1]" }

sub p {
    say Dumper @_;
}

# 31 - input-small.txt answer
# 383 - input-large.txt answer

